### PR TITLE
Introduce very simple A2C implementation

### PR DESCRIPTION
**Pull request by cloderic**
This implementation aims at providing a very simple example of a full `cogment_verse.AgentAdapter` + an example of a synchronous usage of the Run SDK (will parallel rollout of the trial but sequential model updates.

cartpole_simple_a2c
A2C training (trial total reward x trial count) over cartpole with the following hyperparameters

```yaml
simple_a2c_cartpole:
  implementation: simple_a2c_training
  config:
    class_name: data_pb2.SimpleA2CTrainingRunConfig
    environment:
      env_type: "gym"
      env_name: "CartPole-v0"
      seed: 12
    actor:
      num_input: 4
      num_action: 2
    training:
      epoch_count: 100
      epoch_trial_count: 15
      max_parallel_trials: 8
      discount_factor: 0.95
      entropy_coef: 0.01
      value_loss_coef: 0.5
      action_loss_coef: 1.0
      learning_rate: 0.01
    actor_network:
      hidden_size: 64
    critic_network:
      hidden_size: 64
```
Capture d’écran 2021-11-09 à 07 43 16
DQN Training (trial total reward x **steps** count) over cartpole with the following hyperparameters

```yaml
cartpole_dqn:
  implementation: "cogment_verse_run_impl"
  config:
    class_name: data_pb2.RunConfig
    player_count: 1
    epsilon_min: 0.1
    epsilon_steps: 100000
    target_net_update_schedule: 1000
    learning_rate: 1.0e-4
    lr_warmup_steps: 10000
    demonstration_count: 0
    total_trial_count: 10000
    model_publication_interval: 1000
    model_archive_interval_multiplier: 4 # Archive every fourth published model
    render_width: 256
    batch_size: 256
    min_replay_buffer_size: 1000
    max_parallel_trials: 4
    model_kwargs: {}
    max_replay_buffer_size: 100000
    flatten: True
    aggregate_by_actor: False
    framestack: 1
    replay_buffer_config:
      observation_dtype: float32
      action_dtype: int8
    num_input: 4
    num_action: 2
    agent_implementation: dqn
    environment_type: gym
    environment_name: CartPole-v0
```

-----------------

Additional stuffs:
- Support >1 run configurations
- Cleanup some no longer used configuration field
- Support logging full dict / protobuf messages to the experiment tracking

### CODE REVIEW

File: data.proto [hunk 1]
```diff
@@ -126,3 +121,26 @@ message RunConfig {
...
   ReplayBufferConfig replay_buffer_config = 28;
 }
+
+message ModelArchitectureConfig {
+  uint32 hidden_size = 1;
+}
+
+message SimpleA2CTrainingConfig {
+  uint32 epoch_count = 1;
+  uint32 epoch_trial_count = 2;
```
Comment (saikrishna-1996) [inline data.proto:131]: `trial_count` ?

**Review by saikrishna-1996 — DISMISSED** [review]

File: data.proto (hunk 1 above)
Comment (cloderic) [reply data.proto:131]: It's the number of trial per epoch so it made sense to use `epoch_trial_count` I can change it if you think it makes more sense

File: data.proto [hunk 2]
```diff
@@ -126,3 +121,26 @@ message RunConfig {
...
+  uint32 max_parallel_trials = 3;
+  float discount_factor = 4;
+  float entropy_coef = 5;
+  float critic_coef = 6;
```
[the remark this answers is not in the document]
Comment (cloderic) [reply data.proto:135]: If you're positive it's a better name let's go for it (-> can you take a look a the code and check that it's indeed that?)

File: data.proto [hunk 3]
```diff
@@ -126,3 +121,26 @@ message RunConfig {
...
+  float a2c_coef = 7;
+  float learning_rate = 8;
+}
+
+message SimpleA2CTrainingRunConfig {
+  EnvConfig environment = 1;
+  ActorConfig actor = 2;
+  SimpleA2CTrainingConfig training = 3;
+  ModelArchitectureConfig actor_model_architecture = 4;
```
[the remark this answers is not in the document]
Comment (cloderic) [reply data.proto]: Right, I changed the vocabulary from `..._model` to `..._network`

File: data.proto [hunk 4]
```diff
@@ -126,3 +121,26 @@ message RunConfig {
   uint32 framestack = 27;
   ReplayBufferConfig replay_buffer_config = 28;
 }
+
+message NetworkConfig {
```
Comment (jmf1sh) [inline data.proto:125]: Could we use a less generic name than `NetworkConfig`?

**Review by jmf1sh — COMMENTED** [review]
Overall looks really good. I really like the move towards agent-specific run implementations instead of having a single monolithic implementation that tries to do everything.

**Review by jmf1sh — COMMENTED** [review]
Overall looks really good. I really like the move towards agent-specific run implementations instead of having a single monolithic implementation that tries to do everything.

**Review by jmf1sh — APPROVED** [review]
